src/main/java/io/supertokens/oauth/OAuth.java

```diff
@@ -16,40 +16,88 @@
 
-    public static void getAuthorizationUrl(Main main, AppIdentifier appIdentifier, Storage storage, String clientId,
-                                           String redirectURI, String responseType, String scope, String state)
-            throws InvalidConfigException {
+    private static final String LOCATION_HEADER_NAME = "Location";
+    private static final String COOKIES_HEADER_NAME = "Set-Cookie";
+
+    public static OAuthAuthResponse getAuthorizationUrl(Main main, AppIdentifier appIdentifier, Storage storage, String clientId,
+                                                        String redirectURI, String responseType, String scope, String state)
+            throws InvalidConfigException, HttpResponseException, IOException, OAuthException, StorageQueryException,
+            TenantOrAppNotFoundException {
         // TODO:
         // - validate that client_id is present for this tenant
-        // - call hydra
-        // - if location header is:
-        //     - localhost:3000, then we redirect to apiDomain
-        //     - public url for hydra, then we throw a 400 error with the right json
-        //     - else we redirect back to the client
 
         OAuthStorage oauthStorage = StorageUtils.getOAuthStorage(storage);
 
         String redirectTo = null;
+        List<String> cookies = null;
+
+        String publicOAuthProviderServiceUrl = Config.getBaseConfig(main).getOAuthProviderPublicServiceUrl();
 
         if (!oauthStorage.doesClientIdExistForThisApp(appIdentifier, clientId)) {
-            redirectTo = Config.getBaseConfig(main).getOAuthProviderPublicServiceUrl() +
+            redirectTo =  publicOAuthProviderServiceUrl +
                     "/oauth2/fallbacks/error?error=invalid_client&error_description=Client+authentication+failed+%28e" +
                     ".g.%2C+unknown+client%2C+no+client+authentication+included%2C+or+unsupported+authentication" +
                     "+method%29.+The+requested+OAuth+2.0+Client+does+not+exist.";
         } else {
             // we query hydra
+            Map<String, String> queryParamsForHydra = constructHydraRequestParams(clientId, redirectURI, responseType, scope, state);
+            Map<String, String> responseHeaders = new HashMap<>();
+            HttpRequest.sendGETRequestWithResponseHeaders(main, null, Config.getBaseConfig(main).getOAuthProviderPublicServiceUrl(), queryParamsForHydra, 20, 400, null, responseHeaders); // TODO is there some kind of config for the timeouts?
+
+            if(!responseHeaders.isEmpty() && responseHeaders.containsKey(LOCATION_HEADER_NAME)) {
+                String locationHeaderValue = responseHeaders.get(LOCATION_HEADER_NAME);
+
+                if (locationHeaderValue.equals(publicOAuthProviderServiceUrl)){
+                    throw new OAuthException();
+                }
+
+                if (locationHeaderValue.equals("localhost:3000")) {
+                    redirectTo = Multitenancy.getAPIDomain(storage, appIdentifier);
```

Comment:
I think there is a miscommunication here. We just want the redirectTo string to have the domain of `{apiDomain}`. Like the literal string `{apiDomain}/...` and not the value of apiDomain.

Reply:
I feel like I still miss something. Where does this `localhost:3000` come from? This - for now - looks like some magic address. 
I suspect from the example that the `redirectTo` should not be just `{apiDomain}` here. What's the continuation? Is the task here is to change the "localhost:3000" to "{apiDomain}" in the locationHeaderValue?

Reply:
What you have implemented is correct. `localhost:3000` comes from ory (some config in ory is set to that by default). I don't quite recall what it implies. Maybe ask @porcellus. Let me know too :)

Reply:
This localhost/127.0.0.1 comes from a hydra setting. This is the start of the url for the consent/login/etc hydra endpoints, whis is where you would get the redirect after the auth.
We want to change to apiDomain, so hydra and core doesn't have to worry about where to redirect, the sdks will take care of that. 
I didn't change mine, so my hydra instance sends me to 127.0.0.1, @porcellus set his to localhost. This can differ from setup-to-setup, so I propose making it a config in core.